qiskit/transpiler/passes/synthesis/qubit_tracker.py

```diff
@@ -0,0 +1,123 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A qubit state tracker for synthesizing operations with auxiliary qubits."""
+
+from __future__ import annotations
+from collections.abc import Iterable
+from dataclasses import dataclass
+
+
+@dataclass
+class QubitTracker:
+    """Track qubits per index as and their state.
+
+    The states are distinguished into clean (meaning in state :math:`|0\rangle`) or dirty (an
+    unkown state).
+    """
+
+    # This could in future be extended to track different state types, if necessary.
+    # However, using sets of integers here is much faster than e.g. storing a dictionary with
+    # {index: state} entries.
+    qubits: tuple[int]
+    clean: set[int]
+    dirty: set[int]
+
+    def num_clean(self, active_qubits: Iterable[int] | None = None):
+        """Return the number of clean qubits, not considering the active qubits."""
+        # this could be cached if getting the set length becomes a performance bottleneck
+        return len(self.clean.difference(active_qubits or set()))
+
+    def num_dirty(self, active_qubits: Iterable[int] | None = None):
+        """Return the number of dirty qubits, not considering the active qubits."""
+        return len(self.dirty.difference(active_qubits or set()))
+
+    def borrow(self, num_qubits: int, active_qubits: Iterable[int] | None = None) -> list[int]:
+        """Get ``num_qubits`` qubits, excluding ``active_qubits``."""
+        if active_qubits is None:
+            active_qubits = []
+
+        if num_qubits > (available := len(self.qubits) - len(active_qubits)):
+            raise RuntimeError(f"Cannot borrow {num_qubits} qubits, only {available} available.")
+
+        return [qubit for qubit in self.qubits if qubit not in active_qubits][:num_qubits]
+
+    def used(self, qubits: Iterable[int], check: bool = True) -> None:
+        """Set the state of ``qubits`` to used (i.e. False)."""
+        if check:
+            if len(untracked := set(qubits).difference(self.qubits)) > 0:
+                raise ValueError(
+                    "Setting state of untracked qubits:", untracked, "\nTracker state:", self
+                )
+
+        for qubit in qubits:
+            self.clean.discard(qubit)
+
+        self.dirty.update(qubits)
```

Comment:
Nice, this works well, but I think we can optimized this a bit further by converting `qubits` to a set once, which improves the efficiency of set operations. Using set operations (`-=` and `|=`) directly on `self.clean` and `self.dirty` is more efficient than looping through the qubits and performing individual operations.

```suggestion
        qubits = set(qubits)

        if check:
            untracked = qubits - set(self.qubits)
            if untracked:
                raise ValueError(
                    "Setting state of untracked qubits:", untracked, "\nTracker state:", self
                )

        self.clean -= qubits
        self.dirty |= qubits
```

Reply:
Ah great, I was looking for something like `set.update` for removal but didn't know that we can just use `-=`. Thanks!